src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -855,13 +858,19 @@ private ValueTask<FlushResult> TimeFlushUnsynchronizedAsync()
         return _flusher.FlushAsync(_minResponseDataRate, bytesWritten);
     }
 
-    private (long, long) ConsumeConnectionWindow(long bytes)
+    private long CheckConnectionWindow(long bytes)
+    {
+        lock (_windowUpdateLock)
+        {
+            return Math.Min(bytes, _connectionWindow);
```

Comment:
You could still consume the connection window here and avoid retaking the lock, right?

Reply:
I rather not update this state outside of the lock ever 😄. Maybe in preview5.

Reply:
I'm suggesting just going back to the old version of ConsumeConnectionWindow more or less since you always consume what's returned from here (since you've already checked the stream window).

```
private long ConsumeConnectionWindow(long bytes)
{
    lock (_windowUpdateLock)
    {
        var actual = Math.Min(bytes, _connectionWindow);
        _connectionWindow -= actual;
        return actual;
    }
}

Reply:
I see, initially I had this but I decided to keep both consistent.